### DISCUSSION TITLE
URL and UID for  Calendars as well as Events?

### DISCUSSION THREAD

**Discussion by colinmackinlay** [spatie/icalendar-generator#58] [Q&A]
I would like to stream a calendar from Laravel so that calendar clients not only display it but update with any changes; that seems to be the purpose of the `refreshInterval()` method.

The clients need to know where to get the refreshed calendar from which is what I understood the purpose of the `URL` property in https://datatracker.ietf.org/doc/html/rfc5545#section-3.8.4.6 to be.

But `url()` is only implemented for Events. Is there a reason why not and is there another way to get the client to know where to refresh from?

Perhaps some clients record the URL they set the calendar up from so know themselves where to return to but Outlook doesn't seem to do this. Maybe it's because I'm not correctly streaming from Laravel but I'm using the supplied code example 
```$calendar = Calendar::create('Laracon Online');

return response($calendar->get())
    ->header('Content-Type', 'text/calendar')
    ->header('charset', 'utf-8');
```
and I get a downloaded ics file which I then have to open from local file storage.

On a related note, RFC7986 extended the definition of UID to include calendar objects as well as components https://datatracker.ietf.org/doc/html/rfc7986#section-5.3 Is this worth implementing too - it seems like a relativelysimple modification to add say `calendarUniqueIdentifier()` and to give each calendar an initial UID in the same way as for events if this method isn't called.

**Comment by colinmackinlay** [comment]
Of course both URL and UID can be added to a Calendar manually thus:
```
        ->appendProperty(
            TextProperty::create('URL', 'webcals://mydomain/url_for_calendar')
        )
        ->appendProperty(
            TextProperty::create('UID', 'some_unique_id')
        )
```
Outlook will happily accept the calendar with or without the URL included and if the link protocol is set to `webcals://` (or `webcal://` I expect works too) then it asks if you want to subscribe.

In summary, to help others, neither URL nor UID seem necessary for the calendar object in order to subscribe (in Outlook) but could be added for completeness. Outlook doesn't seem to do anything with the UID property on the Calendar - it lets you subscribe multiple times. Other clients might respect this and Outlook might in the future but adding a UID doesn't appear to have an obvious benefit.

When setting the URL of the link to the calendar it is worth using the `webcal(s)` protocols explicitly so that the option to subscribe to updates is given. Outlook stores the URL you open whether it is present in the data or not.

**Comment by rubenvanassche** [comment]
Hi @colinmackinlay,

We haven't implemented `URL` on a calendar object since the RFC describes it can only be used on the following components:

>  Conformance:  This property can be specified once in the "VEVENT", "VTODO", "VJOURNAL", or "VFREEBUSY" calendar components.

As for the `UID` property, I would accept a PR adding them to `Calendar` components just like we do when adding a unique identifier to `Event`s with the `uniqueIdentifier` method.
